database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -14,52 +14,137 @@
+
 import com.google.firebase.database.ChildEventListener;
 import com.google.firebase.database.DataSnapshot;
 import com.google.firebase.database.DatabaseError;
 import com.google.firebase.database.Query;
 import com.google.firebase.database.ValueEventListener;
 
 import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
 import java.util.List;
+import java.util.ListIterator;
+import java.util.concurrent.CopyOnWriteArrayList;
 
 /**
- * This class implements an array-like collection on top of a Firebase location.
+ * This class implements a collection on top of a Firebase location.
  */
-class FirebaseArray implements ChildEventListener, ValueEventListener {
+public class FirebaseArray extends ImmutableList<DataSnapshot> implements ChildEventListener, ValueEventListener {
+    private final List<ChangeEventListener> mListeners = new CopyOnWriteArrayList<>();
+    private List<SubscriptionEventListener> mSubscribers = new ArrayList<>();
+
     private Query mQuery;
-    private ChangeEventListener mListener;
+    private boolean mNotifyListeners = true;
     private List<DataSnapshot> mSnapshots = new ArrayList<>();
 
-    public FirebaseArray(Query ref) {
-        mQuery = ref;
-        mQuery.addChildEventListener(this);
-        mQuery.addValueEventListener(this);
+    /**
+     * @param query The Firebase location to watch for data changes. Can also be a slice of a
+     *              location, using some combination of {@code limit()}, {@code startAt()}, and
+     *              {@code endAt()}.
+     */
+    public FirebaseArray(Query query) {
+        mQuery = query;
     }
 
-    public void cleanup() {
-        mQuery.removeEventListener((ValueEventListener) this);
-        mQuery.removeEventListener((ChildEventListener) this);
+    private static void checkNotNull(Object o) {
+        if (o == null) throw new IllegalArgumentException("Listener cannot be null.");
     }
 
-    public int getCount() {
-        return mSnapshots.size();
+    /**
+     * Add a listener for change events and errors occurring at the location provided in {@link
+     * #FirebaseArray(Query)}.
+     *
+     * @param listener the listener to be called with changes
+     * @return a reference to the listener provided. Save this to remove the listener later
+     * @throws IllegalArgumentException if the listener is null
+     */
+    public ChangeEventListener addChangeEventListener(@NonNull ChangeEventListener listener) {
+        checkNotNull(listener);
+
+        mListeners.add(listener);
+        notifySubscriptionEventListeners(SubscriptionEventListener.EventType.ADDED);
+        if (mListeners.size() <= 1) { // Only start listening when the first listener is added
+            mQuery.addChildEventListener(this);
+            mQuery.addValueEventListener(this);
+        } else {
+            for (int i = 0; i < size(); i++) {
+                listener.onChildChanged(ChangeEventListener.EventType.ADDED, i, -1);
```

Comment:
Should this be `-1` or `i` since the new children are presumably added to the end?